numpy/core/src/multiarray/methods.c

```diff
@@ -823,6 +825,61 @@ array_astype(PyArrayObject *self, PyObject *args, PyObject *kwds)
             return NULL;
         }
 
+        /*
+         * after casting from NPY_DATETIME_NAT to float64, we go back
+         * and replace the spurious floating point values with NPY_NAN
+         * at the appropriate indices; this may also be possible to
+         * achieve upstream, but the PyArray_CopyInto() operation above
+         * uses very low level general array iteration loops where it
+         * may not be desirable to perform a check for NAT on each
+         * iteration
+         */
+
+        /* confine the NaT to NaN substitution checks by type cast */
+        if ((PyArray_DESCR(self)->type_num == NPY_TIMEDELTA) &&
+            ((PyArray_DESCR(ret)->type_num == NPY_FLOAT16) ||
+            (PyArray_DESCR(ret)->type_num == NPY_FLOAT32)  ||
+            (PyArray_DESCR(ret)->type_num == NPY_FLOAT64)  ||
+#ifdef NPY_FLOAT80
+            (PyArray_DESCR(ret)->type_num == NPY_FLOAT80)  ||
+#endif
+#ifdef NPY_FLOAT96
+            (PyArray_DESCR(ret)->type_num == NPY_FLOAT96)  ||
+#endif
+#ifdef NPY_FLOAT128
+            (PyArray_DESCR(ret)->type_num == NPY_FLOAT128) ||
+#endif
+#ifdef NPY_FLOAT256
+            (PyArray_DESCR(ret)->type_num == NPY_FLOAT256)
```

Comment:
Just use `NPY_HALF`, `NPY_FLOAT`, `NPY_DOUBLE`, `NPY_LONGDOUBLE` here, or whatever the names really are. Never used the sized types when writing type-generic code.

Reply:
Is it safe to dispense with the preprocessor guards if I do that?

Reply:
Yes, I think it is - I think long double is always defined, even if its just the same size as double.